### ISSUE TITLE
Stack Overflow with invokedynamic.yield turned on

### ISSUE THREAD

**Issue by pcarlisle** [jruby/jruby#6260]
Since upgrading to 9.2.11.1 we have started to get occasional stack overflow exceptions. I think at this point that turning off invokedynamic.yield has fixed it for us.

Initially the stack traces looked [unremarkable](https://gist.github.com/pcarlisle/d5616999a14b21233e0a53bb59822a45) and certainly not like they should overflow the stack (as in #6124 ), but if I run java with `-XX:+UnlockDiagnosticVMOptions and -XX:+ShowHiddenFrames` I get this: https://gist.github.com/pcarlisle/93f0aebcb90335fd38dbe071d22e22a0

It's cut off by the default stack limit of 1024 but is almost entirely just these two lines repeating:
```
at java.base/java.lang.invoke.DelegatingMethodHandle$Holder.delegate(DelegatingMethodHandle$Holder)
at java.base/java.lang.invoke.LambdaForm$MH/0x00000008418fd040.guard(LambdaForm$MH)
```

A heap dump taken shortly after one of these exceptions also shows hundreds of megabytes retained in a single YieldSite instance.

It takes some hours or even a couple days for these to occur and we no longer have it running this way in production where it was happening but I can try to track down any more information you might need.

(our own tracking is at https://tickets.puppetlabs.com/browse/SERVER-2793 and is public) 

**Environment Information**
  
jruby 9.2.11.1
Linux 3.10

**Comment by headius** [comment]
Nice find! I will look into this.

**Comment by JohnPhillips31416** [comment]
We just hit this issue at Looker as well, will try working around it via invokedynamic.yield=false

**Comment by headius** [comment]
@JohnPhillips31416 That would be my recommendation as well. I guess I never got back to this so I'll have a look now.

**Comment by headius** [comment]
Ok, at a glance I do not see the problem. I need someone to reproduce it with the following settings in place:

* `-Xinvokedynamic.log.binding` passed to JRuby or the equivalent JVM property set. This will output every time a yield gets bound and hopefully show us why we are getting such a long chain (it should give up after two different blocks are seen).
* JVM configured for longer stack traces so I can see the levels before the overflowing recursion. The JVM flag would be `-XX:MaxJavaStackTraceDepth=-1` but if the -1 is rejected just set it to a suitable large size (default is 1024).

Disabling invokedynamic yields altogether should be an effective workaround.

**Comment by headius** [comment]
Oh, if someone can isolate the exact code where this yield is failing, that would also be helpful. The log should provide information about the passed-in block but does not currently show the yield location.

**Comment by headius** [comment]
@JohnPhillips31416 Any updates here? I have not been able to find the issue just staring at the code. Invokedynamic-based yields are fairly limited in how deep they will get.

**Comment by headius** [comment]
I will have to punt this to 9.2.15 because we have not had any follow-up from reporters and I do not know how to reproduce this.

**Comment by JohnPhillips31416** [comment]
My suspicion is that something changed which made stack depths grow, but it's not necessarily a bug with unbounded recursion.

**Comment by headius** [comment]
@ jonathanswenson In #6020 you mention that you had to turn off invokedynamic yield due to this stack overflow issue. Can you help me get a reproduction of that error? I really want to fix it but so far nobody has been able to show me how to trigger it and eyeballing the code has led nowhere.

**Comment by jonathanswenson** [comment]
@headius unfortunately, we've been unable to get a reliable repro for this as well. It seemed to happen after a couple of days of execution suddenly, and the only thing we could do was force a restart. We (I work with @JohnPhillips31416) never really got to the bottom of it, unfortunately. 

We hypothesized it had to do some with deeply nested yielding (effectively auto closing of resources and wrapping methods that would call and yield) as it almost always threw exceptions seemed to be in our deepest nested calls, however it seemed as though we just may have hit some threshold there. 

We attempted to first double the stack size (from 1M that we were using to 2M) but the issue still occurred. According to the stack traces that we logged there were only around 650 frames: lots of these frames were duplicated (all the ones over 10 occurrences): 
```
  11   org.jruby.runtime.BlockBody.yield(BlockBody.java:106)
  12   org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)
  12   org.jruby.runtime.Block.yield(Block.java:184)
  12   org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:191)
  13   org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:182)
  16   org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:148)
  17   org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:80)
  28   org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:172)
  32   org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)
  38   org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:106)
  43   org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:140)
```
Definitely a lot of frames in our code / library code as well.


Let me know if there is any other information I can provide.

**Comment by headius** [comment]
Oh wow, this one is super silly. Fix coming.

**Comment by headius** [comment]
Well you may laugh or you may cry, but this is fixed now.

The problem was the rebind counter, incremented for every new block seen by the yield. Once more than two are encountered, no more are supposed to be bound into the fast path invocation chain. However those subsequent non-fast paths continued to increment the rebind counter, eventually overflowing the 32-bit signed int. One that happened, the counter was always `<= MAX_REBIND` and would it would extend the invocation chain for every block encountered.

This effectively extended every yield site's call path by however many blocks that yield ever encountered. It was not in itself a stack overflow, but it would increase the stack depth of those yields to such a degree that the would eat up thousands of stack frames.

I have pushed a simple fix for this for 9.2.15: only increment the rebind counter when we actually extend the chain.

I will file a separate issue to improve the yield logic to use a more structured tracking mechanism and to be more configurable (i.e. user can extend how many blocks will be allowed to chain).
